File: leetcode/maximum_gap.py
# Given an unsorted array, find the maximum difference between the successive 
# elements in its sorted form.
# Try to solve it in linear time/space.
# Return 0 if the array contains less than 2 elements.
# You may assume all elements in the array are non-negative integers and fit 
# in the 32-bit signed integer range.
# Time Complexity: O(n)
# Space Complexity: O(n)

from math import floor

class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maximumGap(self, nums):
        if not nums or len(nums) == 1: return 0
        lb, width = min(nums), (max(nums) - min(nums)) / (len(nums) - 1.0)
        bucket = [[None, None] for _ in xrange(len(nums))]
        for num in nums:
            tuple = bucket[int(floor((num - lb) / width))]
            tuple[0] = num if not tuple[0] else min(num, tuple[0])
            tuple[1] = num if not tuple[1] else max(num, tuple[1])
        bucket[:] = [tuple for tuple in bucket if tuple[0]]
        return max(bucket[i][0]-bucket[i-1][1] for i in xrange(1, len(bucket)))

if __name__ == "__main__":
    solution = Solution()
    print solution.maximumGap([1, 3, 5, 8, 9])

File: leetcode/maximum_gap.js
/* Given an unsorted array, find the maximum difference between the successive 
 * elements in its sorted form.
 * Try to solve it in linear time/space.
 * Return 0 if the array contains less than 2 elements.
 * You may assume all elements in the array are non-negative integers and fit 
 * in the 32-bit signed integer range.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumGap = function(nums) {
    if (nums.length <= 1)
        return 0;
    var lb = Math.min.apply(Math, nums);
    var width = (Math.max.apply(Math, nums) - lb) / (nums.length - 1);
    if (width === 0)
        return 0;
    for (var bucket = [], i = 0; i < nums.length; ++i)
        bucket.push([null, null]);
    for (var num of nums) {
        var tuple = bucket[Math.floor((num - lb) / width)];
        tuple[0] = tuple[0] === null ? num : Math.min(num, tuple[0]);
        tuple[1] = tuple[1] === null ? num : Math.max(num, tuple[1]);
    }
    var result = [];
    for (tuple of bucket)
        if (tuple[0] !== null)
            result.push(tuple);
    var max = 0;
    for (i = 1; i < result.length; ++i)
        if (result[i][0] - result[i - 1][1] > max)
            max = result[i][0] - result[i - 1][1];
    return max;
};
